### ISSUE TITLE
Stanc3 release for Cmdstan 2.23

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/stanc3#498]
The next Cmdstan release is scheduled for the 20th of April.
For stanc3 that means we will make release binaries for 2.23.

We will once again have a feature freeze period that will start on the 13th of April. That means that no new features should be merged between the 13th and 20th of April.

We currently have 2 non-WIP PRs open for new features: Pedantic mode and `reduce_sum`. 

The closing of this issue will also serve as a mark that normal merging can continue.

I am creating this issue for two reasons:

- as a formal notification to all stanc3 devs that there will be a feature freeze
- to have a central topic for all thing related to the release.

**Comment by rok-cesnovar** [comment]
I am also looking for someone to look at #451 for the Ocaml code review. For the functionalites of the PR we are already testing it rigorously.

**Comment by rok-cesnovar** [comment]
There are also the following issues that should be fixed in time for the release (not necesarilly for the feature freeze). If anyone has time to look at them:
- https://github.com/stan-dev/stanc3/issues/485
- https://github.com/stan-dev/stanc3/issues/497
- https://github.com/stan-dev/stanc3/issues/489
- https://github.com/stan-dev/stanc3/issues/478

**Comment by bbbales2** [comment]
Looks like the language implementation for https://github.com/stan-dev/math/pull/1650 never made it through. The docs did though. We should probably revert the docs.

**Comment by rok-cesnovar** [comment]
Good catch. Its an easy PR to add it to stanc3, but I dont want to rush it and there are plenty of stanc3 bug fix PRs for review already.

**Comment by bbbales2** [comment]
https://github.com/stan-dev/math/pull/1636 did not get language implementations either. Just writing this here so it doesn't get lost.

**Comment by rok-cesnovar** [comment]
Ah yes, I dropped the ball there.

**Comment by bbbales2** [comment]
Here's a draft of the Math release notes.

Let me know if I left anything off or didn't explain something correctly.

----

## Math 3.2.0 Release Notes

### Stan-facing features (available in Stan interfaces):
* Added support for complex numbers at all levels of the autodiff library
* Added new `reduce_sum` utility for parallelizing calculations over multiple cores of one computer
* Sundials upgrade to version 5.2.0
* Support for size 0 matrices [Ben: This needs checked by someone in release candidate, it might be just a Math-facing feature]

### Math-facing features (not available in Stan interfaces):
* discrete_range_lpmf/lcdf/lccdf/rngs/etc. added for uniform discrete random variable
* Added row vector, array and int_array construction utilities
* Adds reverse function for vectors, row vectors, and std::vectors
* Added release notes section to pull request to make assembling release notes easier (this must be filled out for new pulls)
* Added apply operator (standin for C++17 apply)

### Bugfixes:
* Fixed problems with vectorizing neg_binomial_* functions that lead to wrong answers
* Improved lbeta to be more numerically stable with one large and one small argument
* Improved numerical stability with binomial_coefficient_log, neg_binomial_2_lpmf, and neg_binomial_2_log_lpmf computations
* Fixed problem with wrong gradients for large arguments to log_sum_exp
* Fixed bug where normal_id_glm did not work with a sigma is not an autodiff type
* Fixed problem with makefiles on Windows

There were also extensive work in various parts of the Math library adding docs, cleaning up distribution functions, adding tests, generalizing functions to work with special Eigen temporary types, working with new template functions (`return_type_t`, etc.), and otherwise making quality of life improvements.

Thanks to everyone who submitted and reviewed pull requests! The list of pull request submitters for this release are (in no particular order): @t4c1, @mcol, @martinmodrak, @rok-cesnovar, @serban-nicusor-toptal, @SteveBronder, @kedartal, @wds15, @bob-carpenter, @peterwicksstringfield, @andrjohns, @pgree, @IvanYashchuk, and @bbbales2.

**Comment by bbbales2** [comment]
Here are a couple other pulls with stuff we need to add to stanc3/the docs: https://github.com/stan-dev/math/pull/1680, https://github.com/stan-dev/math/pull/1716

I wouldn't say it's dropping the ball on this stuff, cause there's no deadline, we just don't want them to get lost.

**Comment by SteveBronder** [comment]
I need to put this somewhere better but here is the script I used for aggregating all the PRs that happened since last release when I did the release notes before

https://gist.github.com/SteveBronder/bdfe99e216e71a12a41832249516f913

**Comment by SteveBronder** [comment]
> Added support for complex numbers at all levels of the autodiff library

Idk if this is a "stan facing" feature yet since users can't access it yet through the stan language itself (though it will be after we add the stuff it needs to the compiler and io). Might be easier to rephrase this as `External` vs. `Internal` API changes or something. But like also whatever ya know

**Comment by bbbales2** [comment]
Good point. Changed

**Comment by rok-cesnovar** [comment]
Stanc3 release notes:
- Added support for reduce_sum
- Fixed generated code bug for user-defined rng functions
- Fixed uncaught exception for multiple indexes for arrays
- Improved ternary-if typechecking
- Fixed the handling of multiple spaces in "transformed data" and "transformed parameters"
- Separated parser warning for declaring a variables name as a keyword
- Switched to using JSON format when generating mock dataset
- Added exhaustive list of keyword tokens to keyword parser errors 
- Added the use of matrix_cl overloaded math signatures for GLMs with tilde
- Added using std::pow to the generated model code
- Added specifying include files with < > brackets in addition to quotes
- Added an error on model name that starts with a number or non-underscore symbol
- Fixed the spurious integer warning bug
- Fix the offset without multiplier or multiplier without offset bug
- Removed the order dependency of offset/multiplier and lower/upper
- Fixed code generation for foreach loop iteration over elements of sliced arrays
- Fixed transform_inits for constrained types

**Comment by rok-cesnovar** [comment]
Version released. Looks great. Thanks to all that contributed!

If anyone has anything to add on the topic of the release of stanc3 for 2.23 feel free to post here.
